uproot/_connect/to_pandas.py

```diff
@@ -104,95 +105,82 @@ def futures2df(futures, outputtype, entrystart, entrystop, flatten, flatname):
-                                        fn = flatname(name, nn, tup)
-                                        df[fn] = array[nn][(slice(None),) + tup].flatten()
+                array = future()
 
-                else:
-                    fn = flatname(name, None, ())
-                    df[fn] = list(array.flatten())
+                if starts is None:
+                    starts = array.starts
+                    stops = array.stops
+                array = array.flatten()
+
+                length = len(array)
+
+            else:
+                array = future()
+
+            names.append(name)
+            interpretations.append(interpretation)
+            arrays.append(array)
+
+        if length is None:
+            length = len(arrays[0])
+
+        index = pandas.Index(numpy.arange(length))
+        df = outputtype(index=index)
+
+        for name, interpretation, array in zip(names, interpretations, arrays):
+
+            if isinstance(interpretation, asdtype):
+
+                if len(array) < length:
+                    # Invoke jagged broadcasting to align arrays
+                    content = awkward.numpy.zeros(stops.max(), dtype=array.dtype)
+                    array = (awkward.JaggedArray(starts, stops, content) + array).flatten()
 
-            elif isinstance(interpretation, asdtype):
-                df = outputtype(index=index)
                 if interpretation.todims == ():
                     if interpretation.todtype.names is None:
                         fn = flatname(name, None, ())
                         df[fn] = array
-                        scalars.append(fn)
                     else:
                         for nn in interpretation.todtype.names:
                             if not nn.startswith(" "):
                                 fn = flatname(name, nn, ())
                                 df[fn] = array[nn]
-                                scalars.append(fn)
                 else:
                     for tup in itertools.product(*[range(x) for x in interpretation.todims]):
                         if interpretation.todtype.names is None:
                             fn = flatname(name, None, tup)
                             df[fn] = array[(slice(None),) + tup]
-                            scalars.append(fn)
                         else:
                             for nn in interpretation.todtype.names:
                                 if not nn.startswith(" "):
                                     fn = flatname(name, nn, tup)
                                     df[fn] = array[nn][(slice(None),) + tup]
-                                    scalars.append(fn)
 
             else:
                 fn = flatname(name, None, ())
-                df = outputtype(index=index, columns=[fn], data={fn: list(array)})
-                scalars.append(fn)
-
-            out = pandas.merge(out, df, how="outer", left_index=True, right_index=True)
+                df[df] = array
```

Comment:
For the scalars I think the procedure with the `fillna` is not needed, since this operation broadcasts them over the whole column.

Reply:
Oh, that's right. `fillna` was in response to a user request, but it makes the final table look more like a broadcasted frame, anyway. So that's great: another expensive operation doesn't need to be there.